tests/templates/test_subroutines/test_prepselprep.py

```diff
@@ -0,0 +1,265 @@
+            qml.matrix(prepselprep, wire_order=[0, 1, 2])(lcu, control=0),
+        )
+
+    def test_decomposition_negative_coefficients(self):
+        """Test on an LCU with negative coefficients"""
+        dev = qml.device("default.qubit")
+        manual = qml.QNode(self.manual_circuit, dev)
+        prepselprep = qml.QNode(self.prepselprep_circuit, dev)
+
+        lcu = qml.dot([-0.25, -0.75], [qml.Z(2), qml.X(1) @ qml.X(2)])
+        assert np.array_equal(
+            qml.matrix(manual, wire_order=[0, 1, 2])(lcu, control=0),
+            qml.matrix(prepselprep, wire_order=[0, 1, 2])(lcu, control=0),
+        )
+
+    def test_decomposition_complex_coefficients(self):
+        """Test on an LCU with complex coefficients"""
+        dev = qml.device("default.qubit")
+        manual = qml.QNode(self.manual_circuit, dev)
+        prepselprep = qml.QNode(self.prepselprep_circuit, dev)
+
+        lcu = qml.dot([1 + 0.25j, 0 - 0.75j], [qml.Z(2), qml.X(1) @ qml.X(2)])
+        assert np.array_equal(
+            qml.matrix(manual, wire_order=[0, 1, 2])(lcu, control=0),
+            qml.matrix(prepselprep, wire_order=[0, 1, 2])(lcu, control=0),
+        )
+
+    def test_copy(self):
+        """Test the copy function"""
+
+        lcu = qml.dot([1 / 2, 1 / 2], [qml.Identity(1), qml.PauliZ(1)])
+        op = qml.PrepSelPrep(lcu, control=0)
+        op_copy = copy.copy(op)
+
+        assert qml.equal(op, op_copy)
+
+    def test_flatten_unflatten(self):
+        """Test that the class can be correctly flattened and unflattened"""
+
+        lcu = qml.ops.LinearCombination([1 / 2, 1 / 2], [qml.Identity(1), qml.PauliZ(1)])
+        lcu_coeffs, lcu_ops = lcu.terms()
+
+        op = qml.PrepSelPrep(lcu, control=0)
+        data, metadata = op._flatten()
+
+        data_coeffs = [term.terms()[0][0] for term in data]
+        data_ops = [term.terms()[1][0] for term in data]
+
+        assert hash(metadata)
+
+        assert len(data) == len(lcu)
+        assert all(coeff1 == coeff2 for coeff1, coeff2 in zip(lcu_coeffs, data_coeffs))
+        assert all(op1 == op2 for op1, op2 in zip(lcu_ops, data_ops))
+
+        assert metadata == op.control
+
+        new_op = type(op)._unflatten(*op._flatten())
+        assert op.lcu == new_op.lcu
+        assert all(coeff1 == coeff2 for coeff1, coeff2 in zip(op.coeffs, new_op.coeffs))
+        assert all(qml.equal(op1, op2) for op1, op2 in zip(op.ops, new_op.ops))
+        assert op.control == new_op.control
+        assert op.wires == new_op.wires
+        assert op.target_wires == new_op.target_wires
+        assert op is not new_op
+
+
+def test_control_in_ops():
+    """Test that using an operation wire as a control wire results in an error"""
+
+    lcu = qml.dot([1 / 2, 1 / 2], [qml.Identity(0), qml.PauliZ(0)])
+    with pytest.raises(ValueError, match="Control wires should be different from operation wires."):
+        qml.PrepSelPrep(lcu, control=0)
+
```

Comment:
A very interesting test in this subroutine is that `PrepSelPrep` performs a BlockEncoding of the Hamiltonian.

Could you check that this is true for `lcu = qml.dot([0.5, -0.5], [qml.Z(1), qml.X(1)])`?

Reply:
Do you mean another test where I test the template against the a circuit built from the StatePrep and Select templates? Like this

```
 assert np.array_equal(
    qml.matrix(manual, wire_order=[0, 1, 2])(lcu, control=0),
    qml.matrix(prepselprep, wire_order=[0, 1, 2])(lcu, control=0),
)
```

or did you have something else in mind?

Reply:
Yes, I'm suggesting another test.
You could run only your template and check the circuit matrix.
If it works correctly, the top left part of that matrix should match the lcu matrix you are entering as input

Reply:
Could you clarify what you mean by the top left part of the matrix matching the LCU? My implementation gives the same results as in this tutorial https://pennylane.ai/qml/demos/tutorial_lcu_blockencoding/. However, in the tutorial the top left of the block-encoded projector is not equal to the input LCU.

Reply:
Thank you for pointing this out. 
This is due to a small error in the code presented in the demo. 
[This PR](https://github.com/PennyLaneAI/qml/pull/1122) will fix the issue. 
You could check that indeed the block encoding previously commented is fulfilled.

Reply:
I fixed the template so it correctly block-encodes LCUs with both negative and complex coefficients. I added your example, and a couple others, as tests.